File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/history/ui/HistoryFragment.java
package org.mifos.mobilewallet.mifospay.history.ui;

import static org.mifos.mobilewallet.core.domain.model.TransactionType.CREDIT;
import static org.mifos.mobilewallet.core.domain.model.TransactionType.DEBIT;

import android.content.res.Resources;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.google.android.material.chip.Chip;

import org.mifos.mobilewallet.core.domain.model.Transaction;
import org.mifos.mobilewallet.mifospay.R;
import org.mifos.mobilewallet.mifospay.base.BaseActivity;
import org.mifos.mobilewallet.mifospay.base.BaseFragment;
import org.mifos.mobilewallet.mifospay.history.HistoryContract;
import org.mifos.mobilewallet.mifospay.history.presenter.HistoryPresenter;
import org.mifos.mobilewallet.mifospay.history.ui.adapter.HistoryAdapter;
import org.mifos.mobilewallet.mifospay.utils.Constants;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.core.content.res.ResourcesCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.transition.TransitionManager;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class HistoryFragment extends BaseFragment
        implements HistoryContract.HistoryView {

    HistoryAdapter mHistoryAdapter;

    @Inject
    HistoryPresenter mPresenter;
    HistoryContract.TransactionsHistoryPresenter mTransactionsHistoryPresenter;

    @BindView(R.id.cc_history_container)
    ViewGroup historyContainer;

    @BindView(R.id.inc_state_view)
    View vStateView;

    @BindView(R.id.iv_empty_no_transaction_history)
    ImageView ivTransactionsStateIcon;

    @BindView(R.id.tv_empty_no_transaction_history_title)
    TextView tvTransactionsStateTitle;

    @BindView(R.id.tv_empty_no_transaction_history_subtitle)
    TextView tvTransactionsStateSubtitle;

    @BindView(R.id.rv_history)
    RecyclerView rvHistory;

    @BindView(R.id.pb_history)
    ProgressBar pbHistory;

    @BindView(R.id.ll_filter_options)
    LinearLayout filterLayout;

    @BindView(R.id.btn_filter_credits)
    Chip btnFilterCredits;

    @BindView(R.id.btn_filter_all)
    Chip btnFilterAll;

    @BindView(R.id.btn_filter_debits)
    Chip btnFilterDebits;



    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_history, container, false);
        ButterKnife.bind(this, rootView);

        mHistoryAdapter = new HistoryAdapter(position -> mPresenter.handleTransactionClick(position));
        mPresenter.attachView(this);

        setupUi();
        mPresenter.fetchTransactions();

        return rootView;
    }

    private void setupUi() {
        setupSwipeRefreshLayout();
        setupRecyclerView();
        vStateView.setVisibility(View.GONE);
    }

    private void setupRecyclerView() {
        if (getActivity() != null) {
            mHistoryAdapter.setContext(getActivity());
        }
        rvHistory.setLayoutManager(new LinearLayoutManager(getContext()));
        rvHistory.setAdapter(mHistoryAdapter);
    }

    @Override
    public void showTransactionDetailDialog(int transactionIndex, String accountNumber) {
        if (getActivity() != null) {
            TransactionDetailDialog transactionDetailDialog = new TransactionDetailDialog();
            ArrayList<Transaction> transactions = mHistoryAdapter.getTransactions();

            Bundle arg = new Bundle();
            arg.putParcelableArrayList(Constants.TRANSACTIONS, transactions);
            arg.putParcelable(Constants.TRANSACTION, transactions.get(transactionIndex));
            arg.putString(Constants.ACCOUNT_NUMBER, accountNumber);
            transactionDetailDialog.setArguments(arg);

            transactionDetailDialog.show(getActivity().getSupportFragmentManager(),
                    Constants.TRANSACTION_DETAILS);
        }
    }

    private void setupSwipeRefreshLayout() {
        setSwipeEnabled(true);
        getSwipeRefreshLayout().setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                getSwipeRefreshLayout().setRefreshing(false);
                mPresenter.fetchTransactions();
                displayAll();
            }
        });
    }

    @Override
    public void setPresenter(HistoryContract.TransactionsHistoryPresenter presenter) {
        mTransactionsHistoryPresenter = presenter;
    }

    @Override
    public void showStateView(int drawable, int title, int subtitle) {
        TransitionManager.beginDelayedTransition(historyContainer);
        rvHistory.setVisibility(View.GONE);
        pbHistory.setVisibility(View.GONE);
        filterLayout.setVisibility(View.GONE);
        vStateView.setVisibility(View.VISIBLE);
        if (getActivity() != null) {
            Resources res = getResources();
            ivTransactionsStateIcon
                    .setImageDrawable(ResourcesCompat.getDrawable(res, drawable,
                            requireActivity().getTheme()));
            tvTransactionsStateTitle
                    .setText(res.getString(title));
            tvTransactionsStateSubtitle
                    .setText(res.getString(subtitle));
        }
    }

    @Override
    public void showTransactions(List<Transaction> transactions) {
        showRecyclerView();
        mHistoryAdapter.setData(transactions);
    }

    @Override
    public void refreshTransactions(List<Transaction> newTransactions) {
        mHistoryAdapter.setData(newTransactions);
    }

    @Override
    public void showRecyclerView() {
        TransitionManager.beginDelayedTransition(historyContainer);
        vStateView.setVisibility(View.GONE);
        pbHistory.setVisibility(View.GONE);
        rvHistory.setVisibility(View.VISIBLE);
        filterLayout.setVisibility(View.VISIBLE);
    }

    @Override
    public void showHistoryFetchingProgress() {
        TransitionManager.beginDelayedTransition(historyContainer);
        vStateView.setVisibility(View.GONE);
        rvHistory.setVisibility(View.GONE);
        filterLayout.setVisibility(View.GONE);
        pbHistory.setVisibility(View.VISIBLE);
    }

    @OnClick(R.id.btn_filter_all)
    void displayAll() {
        btnFilterAll.setFocusable(true);
        btnFilterAll.setChipBackgroundColorResource(R.color.clickedblue);
        btnFilterCredits.setChipBackgroundColorResource(R.color.changedBackgroundColour);
        btnFilterDebits.setChipBackgroundColorResource(R.color.changedBackgroundColour);
        mTransactionsHistoryPresenter.filterTransactionType(null);
    }

    @OnClick(R.id.btn_filter_credits)
    void displayCredits() {
        btnFilterCredits.setFocusable(true);
        btnFilterCredits.setChipBackgroundColorResource(R.color.clickedblue);
        btnFilterAll.setChipBackgroundColorResource(R.color.changedBackgroundColour);
        btnFilterDebits.setChipBackgroundColorResource(R.color.changedBackgroundColour);
        mTransactionsHistoryPresenter.filterTransactionType(CREDIT);

    }
    @OnClick(R.id.btn_filter_debits)
    void displayDebits() {
        btnFilterDebits.setFocusable(true);
        btnFilterDebits.setChipBackgroundColorResource(R.color.clickedblue);
        btnFilterAll.setChipBackgroundColorResource(R.color.changedBackgroundColour);
        btnFilterCredits.setChipBackgroundColorResource(R.color.changedBackgroundColour);
        mTransactionsHistoryPresenter.filterTransactionType(DEBIT);
    }


}

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/history/ui/HistoryFragment.kt
package org.mifos.mobilewallet.mifospay.history.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.transition.TransitionManager
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.google.android.material.chip.Chip
import dagger.hilt.android.AndroidEntryPoint
import org.mifos.mobilewallet.core.domain.model.Transaction
import org.mifos.mobilewallet.core.domain.model.TransactionType
import org.mifos.mobilewallet.mifospay.R
import org.mifos.mobilewallet.mifospay.base.BaseFragment
import org.mifos.mobilewallet.mifospay.history.HistoryContract.HistoryView
import org.mifos.mobilewallet.mifospay.history.HistoryContract.TransactionsHistoryPresenter
import org.mifos.mobilewallet.mifospay.history.presenter.HistoryPresenter
import org.mifos.mobilewallet.mifospay.history.ui.adapter.HistoryAdapter
import org.mifos.mobilewallet.mifospay.utils.Constants
import org.mifos.mobilewallet.mifospay.utils.RecyclerItemClickListener
import javax.inject.Inject

@AndroidEntryPoint
class HistoryFragment : BaseFragment(), HistoryView {
    @JvmField
    @Inject
    var mHistoryAdapter: HistoryAdapter? = null

    @JvmField
    @Inject
    var mPresenter: HistoryPresenter? = null
    var mTransactionsHistoryPresenter: TransactionsHistoryPresenter? = null

    @JvmField
    @BindView(R.id.cc_history_container)
    var historyContainer: ViewGroup? = null

    @JvmField
    @BindView(R.id.inc_state_view)
    var vStateView: View? = null

    @JvmField
    @BindView(R.id.iv_empty_no_transaction_history)
    var ivTransactionsStateIcon: ImageView? = null

    @JvmField
    @BindView(R.id.tv_empty_no_transaction_history_title)
    var tvTransactionsStateTitle: TextView? = null

    @JvmField
    @BindView(R.id.tv_empty_no_transaction_history_subtitle)
    var tvTransactionsStateSubtitle: TextView? = null

    @JvmField
    @BindView(R.id.rv_history)
    var rvHistory: RecyclerView? = null

    @JvmField
    @BindView(R.id.pb_history)
    var pbHistory: ProgressBar? = null

    @JvmField
    @BindView(R.id.ll_filter_options)
    var filterLayout: LinearLayout? = null

    @JvmField
    @BindView(R.id.btn_filter_credits)
    var btnFilterCredits: Chip? = null

    @JvmField
    @BindView(R.id.btn_filter_all)
    var btnFilterAll: Chip? = null

    @JvmField
    @BindView(R.id.btn_filter_debits)
    var btnFilterDebits: Chip? = null
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val rootView = inflater.inflate(R.layout.fragment_history, container, false)
        ButterKnife.bind(this, rootView)
        mPresenter!!.attachView(this)
        setupUi()
        mPresenter!!.fetchTransactions()
        return rootView
    }

    private fun setupUi() {
        setupSwipeRefreshLayout()
        setupRecyclerView()
        vStateView!!.visibility = View.GONE
    }

    private fun setupRecyclerView() {
        if (activity != null) {
            mHistoryAdapter!!.setContext(activity)
        }
        rvHistory!!.layoutManager = LinearLayoutManager(context)
        rvHistory!!.adapter = mHistoryAdapter
        rvHistory!!.addOnItemTouchListener(
            RecyclerItemClickListener(
                context,
                object : RecyclerItemClickListener.OnItemClickListener {
                    override fun onItemClick(childView: View, position: Int) {
                        mPresenter!!.handleTransactionClick(position)
                    }

                    override fun onItemLongPress(childView: View, position: Int) {}
                })
        )
    }

    private fun setupSwipeRefreshLayout() {
        setSwipeEnabled(true)
        swipeRefreshLayout.setOnRefreshListener {
            swipeRefreshLayout.isRefreshing = false
            mPresenter!!.fetchTransactions()
            displayAll()
        }
    }

    override fun showStateView(drawable: Int, title: Int, subtitle: Int) {
        TransitionManager.beginDelayedTransition(historyContainer!!)
        rvHistory!!.visibility = View.GONE
        pbHistory!!.visibility = View.GONE
        filterLayout!!.visibility = View.GONE
        vStateView!!.visibility = View.VISIBLE
        if (activity != null) {
            val res = resources
            ivTransactionsStateIcon
                ?.setImageDrawable(res.getDrawable(drawable))
            tvTransactionsStateTitle
                ?.setText(res.getString(title))
            tvTransactionsStateSubtitle
                ?.setText(res.getString(subtitle))
        }
    }

    override fun showTransactions(transactions: List<Transaction>?) {
        showRecyclerView()
        mHistoryAdapter!!.setData(transactions)
    }

    override fun showEmptyTransactionTypeStateView(
        drawable: Int,
        title: String?,
        subtitle: String?
    ) {
        TransitionManager.beginDelayedTransition(historyContainer!!)
        rvHistory!!.visibility = View.GONE
        pbHistory!!.visibility = View.GONE
        filterLayout!!.visibility = View.VISIBLE
        vStateView!!.visibility = View.VISIBLE
        if (activity != null) {
            val res = resources
            ivTransactionsStateIcon
                ?.setImageDrawable(res.getDrawable(drawable))
            tvTransactionsStateTitle?.text = title
            tvTransactionsStateSubtitle?.text = subtitle
        }
    }

    override fun showTransactionDetailDialog(transactionIndex: Int, accountNumber: String?) {
        if (activity != null) {
            val transactionDetailDialog = TransactionDetailDialog()
            val transactions = mHistoryAdapter!!.getTransactions()
            val arg = Bundle()
            arg.putParcelableArrayList(Constants.TRANSACTIONS, transactions)
            arg.putParcelable(Constants.TRANSACTION, transactions!![transactionIndex])
            arg.putString(Constants.ACCOUNT_NUMBER, accountNumber)
            transactionDetailDialog.arguments = arg
            transactionDetailDialog.show(
                requireActivity().supportFragmentManager,
                Constants.TRANSACTION_DETAILS
            )
        }
    }

    override fun showRecyclerView() {
        TransitionManager.beginDelayedTransition(historyContainer!!)
        vStateView!!.visibility = View.GONE
        pbHistory!!.visibility = View.GONE
        rvHistory!!.visibility = View.VISIBLE
        filterLayout!!.visibility = View.VISIBLE
    }

    override fun showHistoryFetchingProgress() {
        TransitionManager.beginDelayedTransition(historyContainer!!)
        vStateView!!.visibility = View.GONE
        rvHistory!!.visibility = View.GONE
        filterLayout!!.visibility = View.GONE
        pbHistory!!.visibility = View.VISIBLE
    }

    override fun refreshTransactions(transactions: List<Transaction>?) {
        showRecyclerView()
        mHistoryAdapter!!.setData(transactions)
    }

    override fun setPresenter(presenter: TransactionsHistoryPresenter?) {
        mTransactionsHistoryPresenter = presenter
    }

    @OnClick(R.id.btn_filter_all)
    fun displayAll() {
        btnFilterAll!!.isFocusable = true
        btnFilterAll!!.setChipBackgroundColorResource(R.color.clickedblue)
        btnFilterCredits!!.setChipBackgroundColorResource(R.color.changedBackgroundColour)
        btnFilterDebits!!.setChipBackgroundColorResource(R.color.changedBackgroundColour)
        mTransactionsHistoryPresenter!!.filterTransactionType(null)
    }

    @OnClick(R.id.btn_filter_credits)
    fun displayCredits() {
        btnFilterCredits!!.isFocusable = true
        btnFilterCredits!!.setChipBackgroundColorResource(R.color.clickedblue)
        btnFilterAll!!.setChipBackgroundColorResource(R.color.changedBackgroundColour)
        btnFilterDebits!!.setChipBackgroundColorResource(R.color.changedBackgroundColour)
        mTransactionsHistoryPresenter!!.filterTransactionType(TransactionType.CREDIT)
    }

    @OnClick(R.id.btn_filter_debits)
    fun displayDebits() {
        btnFilterDebits!!.isFocusable = true
        btnFilterDebits!!.setChipBackgroundColorResource(R.color.clickedblue)
        btnFilterAll!!.setChipBackgroundColorResource(R.color.changedBackgroundColour)
        btnFilterCredits!!.setChipBackgroundColorResource(R.color.changedBackgroundColour)
        mTransactionsHistoryPresenter!!.filterTransactionType(TransactionType.DEBIT)
    }
}